### PR TITLE
Use classy prisms for errors

### PR DESCRIPTION
**Pull request by michaelpj**
This is a POC of using [classy prisms](https://gist.github.com/nkpart/c3bcb48c97c5ded6e277) to handle the embedding-errors problem.

Here's my assessment of it.

Pros:
- No `convertErrors` weirdness.
    - Crucially, this means I'm pretty sure this will keep working as the projects grow.
- Very explicit about what errors can be thrown.
- Actual use sites are nice (look at `Plugin.hs`, which uses functions from both `language-plutus-core` and `plutus-ir` that throw different kinds of errors).

Cons:
- Sometimes painfully explicit - we get quite big constraints for some of the `language-plutus-core` functions because they can throw lots of kinds of errors.
- Error types are merely constrained, so if you use the error polymorphically you sometimes now need to specify what the error type is.
    - We need to do the constraint-based version rather than using concrete types because otherwise we lose all the benefit when we try to use the functions in other places.
- Need to write a (trivial) instance for each kind of error you want to lift into your concrete error type.
    - I tried to write "chaining" instances like `instance AsError e a => AsParseError e a`, so that you would only have to say you can embed e.g. `language-plutus-core`'s concrete error type and you would get the others for free, but this immediately gets stuck in a morass of overlapping and undecidable instances, and I gave up.
- Some TH.

Neutral:
- Lens dep. It seems like we were going to get this at some point anyway, and it lets us clear up some code that duplicates what it provides.
- Still doesn't solve the problem of *handling* a specific type of error and having it fall out, but this seems to be even harder (nobody has a good solution to this AFAICT except maybe extensible-effects/freer-simple or something).

### CODE REVIEW

File: language-plutus-core/src/Language/PlutusCore.hs [hunk 1]
```diff
@@ -141,47 +143,54 @@ fileType :: FilePath -> IO T.Text
 fileType = fileNormalizeType False
 
 fileNormalizeType :: Bool -> FilePath -> IO T.Text
-fileNormalizeType norm = fmap (either prettyPlcDefText id . printNormalizeType norm) . BSL.readFile
+fileNormalizeType norm = fmap (either @(Error AlexPosn) prettyPlcDefText id . printNormalizeType norm) . BSL.readFile
```
Comment (vmchale) [inline language-plutus-core/src/Language/PlutusCore.hs]: This should just be a type signature. No need for type applications here.

File: language-plutus-core/src/Language/PlutusCore/Name.hs [hunk 2]
```diff
@@ -2,8 +2,10 @@
 {-# LANGUAGE DerivingStrategies     #-}
 {-# LANGUAGE FlexibleInstances      #-}
 {-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
 {-# LANGUAGE MultiParamTypeClasses  #-}
 {-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE TypeFamilies           #-}
```
Comment (vmchale) [inline language-plutus-core/src/Language/PlutusCore/Name.hs]: Remove this.

File: language-plutus-core/src/Language/PlutusCore/TH.hs [hunk 3]
```diff
@@ -130,7 +129,7 @@ compileTerm s = do
             quoted :: Quote (Term TyName Name ())
             quoted = do
                 -- See note [Parsing and TH stages]
-                runtimeT <- (fmap void . MM.hoist (Identity . unsafeDropErrors) . parseTerm . strToBs) s
+                runtimeT <- unsafeDropErrors @(Error AlexPosn) <$> (runExceptT $ (fmap void . parseTerm . strToBs) s)
```
Comment (vmchale) [inline language-plutus-core/src/Language/PlutusCore/TH.hs]: Can this be replaced with a type signature?

File: language-plutus-core/test/Spec.hs [hunk 4]
```diff
@@ -253,9 +254,9 @@ testRebindCapturedVariable =
 
 tests :: TestTree
 tests = testCase "example programs" $ fold
-    [ format cfg "(program 0.1.0 [(builtin addInteger) x y])" @?= Right "(program 0.1.0\n  [ [ (builtin addInteger) x ] y ]\n)"
-    , format cfg "(program 0.1.0 doesn't)" @?= Right "(program 0.1.0\n  doesn't\n)"
-    , format cfg "{- program " @?= Left (ParseError (LexErr "Error in nested comment at line 1, column 12"))
+    [ format @(ParseError AlexPosn) cfg "(program 0.1.0 [(builtin addInteger) x y])" @?= Right "(program 0.1.0\n  [ [ (builtin addInteger) x ] y ]\n)"
```
Comment (vmchale) [inline language-plutus-core/test/Spec.hs]: Get rid of the type applications here.

File: language-plutus-core/src/Language/PlutusCore/Name.hs [hunk 5]
```diff
@@ -49,6 +51,7 @@ data Name a = Name { nameAttribute :: a
 newtype TyName a = TyName { unTyName :: Name a }
     deriving (Show, Generic, Lift)
     deriving newtype (Eq, Ord, Functor, NFData)
+instance Wrapped (TyName a)
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Name.hs]: I'm not a fan of this change. Now you need to provide an instance for each newtype forever, that's just boilerplate. And you need to enable either `GADTs` or `TypeFamilies` for this, boilerplate again. And type signatures did not become shorter:

```
-    :: (Newtype new, HasUnique (Unwrap new) unique', Coercible unique' unique)
+    :: (Wrapped new, HasUnique (Unwrapped new) unique', Coercible unique' unique)
```

This type class is only needed for wrapping/unwrapping non-newtypes (I suppose in can handle a non-existential GADT with a single constructor). Do we really need this generality at the cost of writing more boilerplate?

File: plutus-ir/src/Language/PlutusIR/Compiler/Types.hs [hunk 6]
```diff
@@ -19,6 +19,6 @@ type PLCType a = PLC.Type PLC.TyName (Provenance a)
 type PIRTerm a = PIR.Term PIR.TyName PIR.Name (Provenance a)
 type PIRType a = PIR.Type PIR.TyName (Provenance a)
 
-type Compiling m a = (Monad m, MonadReader (Provenance a) m, MonadError (Error (Provenance a)) m, MonadQuote m)
+type Compiling m e a = (Monad m, MonadReader (Provenance a) m, AsError e (Provenance a), MonadError e m, MonadQuote m)
```
Comment (effectfully) [inline plutus-ir/src/Language/PlutusIR/Compiler/Types.hs]: Here is your unused type variable :)
It's a shame we can't turn a functional dependency into a type family and get rid of this `e`.

File: language-plutus-core/src/Language/PlutusCore/Name.hs (hunk 5 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Name.hs]: I don't really think we're writing much boilerplate. We're going to use this maybe two or three times. The main thing I like about this is just that we're no longer using our own code, which just makes me more confident that it's right.

File: plutus-ir/src/Language/PlutusIR/Compiler/Types.hs (hunk 6 above)
Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Compiler/Types.hs]: I probably could instead define my own typeclass that has these constraints as super-constraints and has an associated type, but I'm not sure I want to.

File: language-plutus-core/src/Language/PlutusCore.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore.hs]: You really don't like type applications, do you? :) I find it nicer than lifting random subexpressions, but I'll do that if you prefer.

File: language-plutus-core/src/Language/PlutusCore.hs [hunk 7]
```diff
@@ -141,47 +143,54 @@ fileType :: FilePath -> IO T.Text
...
 
 -- | Parse a program and typecheck it.
-parseTypecheck :: (MonadError (Error AlexPosn) m, MonadQuote m) => Natural -> BSL.ByteString -> m (NormalizedType TyNameWithKind ())
+parseTypecheck
+    :: (AsParseError e AlexPosn,
+        AsNormalizationError e TyName Name AlexPosn,
+        AsRenameError e AlexPosn,
+        AsTypeError e AlexPosn,
+        MonadError e m,
+        MonadQuote m)
```
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore.hs]: I see. Well, that's a pity.

File: language-plutus-core/src/Language/PlutusCore/Name.hs (hunk 5 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Name.hs]: I don't think safe coercions can be wrong at all. Well, I do not hate this change, I just like the previous version more, but if you prefer to use the `lens`' one, that makes sense too.

File: plutus-ir/src/Language/PlutusIR/Compiler/Types.hs (hunk 6 above)
Comment (effectfully) [reply plutus-ir/src/Language/PlutusIR/Compiler/Types.hs]: I wouldn't want as well.

File: language-plutus-core/src/Language/PlutusCore.hs (hunk 1 above)
Comment (mchakravarty) [reply language-plutus-core/src/Language/PlutusCore.hs]: Hmm, I do like the type application TBH. What is bad about it?

Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore.hs]: I have nothing against type applications as well.

**Comment by michaelpj** [comment]
> How about using ConstraintKinds to define synonyms for common combinations?

Yeah, I considered that. They're big, but don't appear that often, because there's only the one typechecking function etc. I think it's okay.

> Or you could use class context wildcards, but I am not sure that I like them much

:see_no_evil:

**Comment by michaelpj** [comment]
I already did the thing where I hoist out subexpressions and give them type signatures, which isn't *too* much worse, so might as well leave it.

**Comment by michaelpj** [comment]
Okay, it seems like nobody hates this, so I propose we actually do it. Any more detailed critique welcome!

File: language-plutus-core/src/Language/PlutusCore/Name.hs (hunk 5 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Name.hs]: We're still using the coercions! The only thing that's not there is the clever thing where you work out that something is a newtype without having to say anything.

Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Name.hs]: > We're still using the coercions!

Sure, I responded to:

> makes me more confident that it's right.

**Comment by michaelpj** [comment]
Here goes.

**Pull request merged**